src/Validation/DNSCheckValidation.php

```diff
@@ -57,21 +89,46 @@ public function getWarnings()
     protected function checkDNS($host)
     {
         $variant = INTL_IDNA_VARIANT_2003;
-        if ( defined('INTL_IDNA_VARIANT_UTS46') ) {
+        if (defined('INTL_IDNA_VARIANT_UTS46')) {
             $variant = INTL_IDNA_VARIANT_UTS46;
         }
         $host = rtrim(idn_to_ascii($host, IDNA_DEFAULT, $variant), '.') . '.';
 
-        $Aresult = true;
-        $MXresult = checkdnsrr($host, 'MX');
+        // Get all MX, A and AAAA DNS records
+        $dnsRecords = dns_get_record($host, DNS_MX + DNS_A + DNS_AAAA);
 
-        if (!$MXresult) {
-            $this->warnings[NoDNSMXRecord::CODE] = new NoDNSMXRecord();
-            $Aresult = checkdnsrr($host, 'A') || checkdnsrr($host, 'AAAA');
-            if (!$Aresult) {
-                $this->error = new NoDNSRecord();
+        // No MX, A or AAAA records
+        if (empty($dnsRecords)) {
+            $this->error = new NoDNSRecord();
+            return false;
+        }
+
+        $aRecords  = [];
+        $mxRecords = [];
+
+        // Iterate over all returned DNS records
+        foreach ($dnsRecords as $dnsRecord) {
```

Comment:
Please extract to a separate method.

Reply:
What is your reasoning for this? It's quite simple, so I feel that splitting it out will make it harder to read. If we weren't returning in it, I would agree, but returning there allows us to potentially save a few cpu cycles.

Reply:
I see your point.
Main reasons are that there are 3 nesting levels, two main decisions branches and there is an important decision in the most inner one. So I'd say saving a couple of cycles is an acceptable price.
What I see is somehiting like
```php
//... L105

return $this->validateDNSRecords ($dnsRecords);
}

private function validateDNSRecords ($dnsRecords) {

    foreach ($dnsRecords as $dnsRecord) {
       if (!$this->validateMXRecord($dnsRecord)) { return false; }
       //add here other validations, i.e: AAA records?
    }
    if (empty($this->mxRecords)) {
            $this->warnings[NoDNSMXRecord::CODE] = new NoDNSMXRecord();
    }
    return true;
}
// returns boolean
private function validateMXRecord($dnsRecord) {
     if ($dnsRecord['type'] != 'MX') { return true; }
    // "Null MX" record indicates the domain accepts no mail (https://tools.ietf.org/html/rfc7505)
    if (empty($dnsRecord['target']) || $dnsRecord['target'] == '.') {
                    $this->error = new DomainAcceptsNoMail();
                    return false;
    }
    $this->mxRecords[] = $dnsRecord;
    return true;
}
```

Reply:
I've split this out locally but I think it makes it significantly harder to understand what is going on at a glance. Ultimately your call though... Kinda feels like you only really want it for the single return in `checkDns()` haha

Reply:
I don't see any changes, have you pushed the commits?
Hahaha, if we could even remove that one return...  just kidding! What I'm looking for is to make it understandable by encapsulating, that ultimately did drive to a single return in the calling method, but as you can see on the private ones I have no issues with multiple returns.

Let see your changes and we can then solve this one!
Thanks!

Reply:
👍